File: python/euler/bin/p0079.py
# project euler: problem 79

#   I saw the Wikipedia page about topological sorting.
#     https://en.wikipedia.org/wiki/Topological_sorting
#
#   Note: This implementation finds only one topological sort not all.

from collections import defaultdict
from collections.abc import Mapping
from functools import reduce
from typing import TextIO


def dfs(graph: Mapping[str, list[str]], perm: list[str], v: str) -> list[str]:
    def visit(temp: list[str], visited: list[str], node: str) -> list[str]:
        if node in temp:
            raise AssertionError('cycle detected')
        if node in visited:
            return visited
        if node in graph:
            acc = visited
            for next_v in graph[node]:
                acc = visit([node] + temp, acc, next_v)
            return [node] + acc
        else:
            return [node]

    return visit([], perm, v)


def compute(fh: TextIO) -> str:
    def parse_data(fh: TextIO) -> dict[str, list[str]]:
        result: dict[str, list[str]] = defaultdict(list)
        for k, v in reduce(
            lambda x, y: x + y,
            [[(x[0], x[1]), (x[1], x[2]), (x[0], x[2])] for x in fh.read().splitlines()],
        ):
            result[k].append(v)

        for k, vs in result.items():
            result[k] = sorted(set(vs))

        return result

    graph = parse_data(fh)
    acc: list[str] = []
    for k in graph:
        acc = dfs(graph, acc, k)

    return reduce(lambda x, y: x + y, acc)


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p079_keylog.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0079.ts
// project euler: problem 79

import { assetData } from "../lib/asset.ts";

const dfs = (
  graph: Map<string, Set<string>>,
  perm: string[],
  v: string,
): string[] => {
  const visit = (temp: string[], visited: string[], node: string): string[] => {
    if (temp.includes(node) === true) {
      throw new Error("cycle path is found");
    }
    if (visited.includes(node) === true) {
      return visited;
    }
    if (graph.has(node) === true) {
      let acc = visited;
      for (const v of graph.get(node)!) {
        acc = visit([node].concat(temp), acc, v);
      }
      return [node].concat(acc);
    } else {
      return [node];
    }
  };

  return visit([], perm, v);
};

const parseData = (data: string): string[][] => {
  const splitLines = (str: string): string[] => {
    const result = str.split(/\r?\n/);
    if (result.at(-1) === "") {
      return result.slice(0, -1);
    } else {
      return result;
    }
  };

  return splitLines(data)
    .map((x) => x.split(""))
    .map((y) => [[y[0], y[1]], [y[0], y[2]], [y[1], y[2]]]).flat();
};

export const compute = (data: string): string => {
  const graph = new Map<string, Set<string>>();
  for (const [k, v] of parseData(data)) {
    graph.set(k, (graph.get(k) || new Set<string>()).add(v));
  }

  let acc: string[] = [];
  for (const v of graph.keys()) {
    acc = dfs(graph, acc, v);
  }

  return acc.join("");
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p079_keylog.txt"));
  return compute(data);
};
